src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -72,95 +95,152 @@ public HdfsWriter(HdfsSinkConnectorConfig connectorConfig) {
     }
   }
 
-
   public Map<TopicPartition, Long> getPreviousOffsets() {
     return offsets;
   }
 
-  public void writeRecord(TopicPartition topicPart, SinkRecord record)
-      throws CopycatException{
+  public void writeRecord(final TopicPartition topicPart, final SinkRecord record)
+      throws CopycatException, InterruptedException {
+    lock.lock();
     try {
+      recover(topicPart);
+
       RecordWriter<Long, SinkRecord> writer = getWriter(topicPart, record);
       writer.write(System.currentTimeMillis(), record);
+
       if (!offsets.containsKey(topicPart)) {
         offsets.put(topicPart, record.getOffset() - 1);
       }
       updateRecordCounter(topicPart);
+
+      if (!tasks.containsKey(topicPart)) {
+        ScheduledFuture task = timedRotateThread.schedule(new Runnable() {
+          @Override
+          public void run() {
+            lock.lock();
+            try {
+              rotate(topicPart);
+            } catch (IOException | InterruptedException e) {
+              throw new CopycatRuntimeException(e);
+            } finally {
+              lock.unlock();
+            }
+          }
+        }, rotateInterval, TimeUnit.SECONDS);
+        tasks.put(topicPart, task);
+      }
       if (shouldRotate(topicPart)) {
         rotate(topicPart);
       }
     } catch (IOException e) {
       throw new CopycatRuntimeException(e);
+    } finally {
+      lock.unlock();
     }
   }
 
-  public void close() throws IOException, CopycatException {
-    for (TopicPartition topicPart: writers.keySet()) {
+  public void close() throws IOException, CopycatException, InterruptedException {
+    timedRotateThread.shutdown();
+
+    for (TopicPartition topicPart : writers.keySet()) {
```

Comment:
Don't we need to acquire the lock to do this? Maybe we need to clarify the semantics of `close()` in copycat -- at this point you can't be guaranteed the thread calling `put()` has exited because the task might have control of it. So `stop()` is responsible for signaling the thread to stop, then making sure it has stopped before doing any additional cleanup.

Would it help if copycat separated this into two steps, something like `stop()` and `cleanup()` such that `cleanup()` is only invoked once the other thread has exited?